velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+          rawBuildIndices_.begin() + (i + 1) * buildSize,
           0);
     }
   }
 
-  auto buildRowVector =
-      buildData_.value()[buildIndex_]->asUnchecked<RowVector>();
-  for (const auto& projection : buildProjections_) {
-    VectorPtr buildVector = buildRowVector->childAt(projection.inputChannel);
-
-    if (buildIndices) {
-      buildVector = BaseVector::wrapInDictionary(
-          BufferPtr(nullptr), buildIndices, size, buildVector);
-    }
-    output->childAt(projection.outputChannel) = buildVector;
-  }
+  projectChildren(output, input_, probeProjections, batchSize, probeIndices_);
+  projectChildren(
+      output,
+      buildData_.value()[buildIndex_],
+      buildProjections,
+      batchSize,
+      buildIndices_);
+  return output;
+}
 
+bool NestedLoopJoinProbe::advanceIndex(vector_size_t probeCnt) {
   probeRow_ += probeCnt;
-  if (probeRow_ == inputSize) {
-    probeRow_ = 0;
+  if (probeRow_ < input_->size()) {
+    return false;
+  }
+  probeRow_ = 0;
+  prevProbeCnt_ = 0;
+  do {
     ++buildIndex_;
-    if (buildIndex_ == buildData_->size()) {
-      buildIndex_ = 0;
-      input_.reset();
+  } while (!hasProbedAllBuildData() &&
+           !buildData_.value()[buildIndex_]->size());
+  return hasProbedAllBuildData();
+}
+
+RowVectorPtr NestedLoopJoinProbe::doMatching(vector_size_t probeCnt) {
+  VELOX_CHECK_NOT_NULL(input_);
+  VELOX_CHECK(!hasProbedAllBuildData());
+
+  if (joinCondition_ == nullptr) {
+    return getCrossProduct(
+        probeCnt, outputType_, identityProjections_, buildProjections_);
+  }
+
+  auto filterInput = getCrossProduct(
+      probeCnt,
+      filterInputType_,
+      filterProbeProjections_,
+      filterBuildProjections_);
+
+  if (filterInputRows_.size() != filterInput->size()) {
+    filterInputRows_.resizeFill(filterInput->size(), true);
+  }
+  VELOX_CHECK(filterInputRows_.isAllSelected());
+
+  std::vector<VectorPtr> filterResult;
+  EvalCtx evalCtx(
+      operatorCtx_->execCtx(), joinCondition_.get(), filterInput.get());
+  joinCondition_->eval(0, 1, true, filterInputRows_, evalCtx, filterResult);
+  DecodedVector decodedFilterResult;
+  decodedFilterResult.decode(*filterResult[0], filterInputRows_);
+
+  const vector_size_t resultSize = decodedFilterResult.size();
+  auto rawProbeOutMapping =
+      initializeRowNumberMapping(probeOutMapping_, resultSize, pool());
+  auto rawBuildOutMapping =
+      initializeRowNumberMapping(buildOutMapping_, resultSize, pool());
+  auto probeIndices = probeIndices_->asMutable<vector_size_t>();
+  auto buildIndices = buildIndices_->asMutable<vector_size_t>();
+  int32_t outputSize{0};
+  for (auto i = 0; i < resultSize; ++i) {
+    if (!decodedFilterResult.isNullAt(i) &&
+        decodedFilterResult.valueAt<bool>(i)) {
+      rawProbeOutMapping[outputSize] = probeIndices[i];
+      rawBuildOutMapping[outputSize] = buildIndices[i];
+      outputSize++;
+    }
+  }
+  if (needsProbeMismatch(joinType_)) {
```

Comment:
Shall we cache this flag above and set probeMatched_ in match check loop above? Otherwise we might set the same probe row many times which is the build vector size?

Reply:
My intention was to reduce branching, cuz in the end, we need to test / set some state for each row in decoded result.
Say we wanna check if certain probe row is matched in match check loop, we may take either of follow method:
1. As for each `buildSize` rows in decoded result, the probe row are the same. We have a bool `currentProbeRowMatched` for current probe row:
```
if needsProbeMismatch && !currentProbeRowMatched
    currentProbeRowMatched = true
```
then update probeMatched if probe row has `currentProbeRowMatched`

2. Have a variable `prevMatchedProbeRow`:
```
if needsProbeMismatch && currentMatchedProbeRow != prevMatchedProbeRow
    prevMatchedProbeRow = currentMatchedProbeRow
    probeMatched.setValid(currentMatchedProbeRow, true)
```

Method 1 tests `currentProbeRowMatched` `decodedResult.size()` times, method 2 tests `prevMatchedProbeRow` `decodedResult.size()` times, and maybe cache locality is also affected ?

Reply:
Sorry, I bit lost of previous version. I am good with the current version. In general, we could optimize the code based on profiling data if the optimized code is more complicated to start with implementation wise.